nearswap/src/twap.rs

```diff
@@ -0,0 +1,601 @@
+                timestamp,
+                price1,
+                price2,
+                MAX_LENGTH
+            );
+        }
+
+        // update mean
+        self.update_mean(self.observations.len());
+    }
+
+    pub fn update_mean(&mut self, len: u64) {
+        self.mean_1min = self.calculate_mean(Mean::M_1MIN, len);
+        self.mean_5min = self.calculate_mean(Mean::M_5MIN, len);
+        self.mean_1h = self.calculate_mean(Mean::M_1H, len);
+        self.mean_12h = self.calculate_mean(Mean::M_12H, len);
+    }
+}
+
+#[derive(BorshSerialize, BorshDeserialize)]
+pub struct Observation {
+    // timestamp
+    pub block_timestamp: u64,
+    // Number of observations till block_timestamp
+    pub num_of_observations: u128,
+    // cumulative price of token1 till block_timestamp
+    pub price1_cumulative: u128,
+    // cumulative price of token2 till block_timestamp
+    pub price2_cumulative: u128,
+}
+
+/**
+Transforms a previous observation into a new observation.
+Parameters:
++ `block_timestamp`: _must_ be chronologically equal to or greater than last.block_timestamp.
++ `last`: The specified observation to be transformed.
++ `price1`: price of first token.
++ `price2`: price of second token.
+*/
+pub fn transform(
+    last: &Observation,
+    block_timestamp: u64,
+    price1: u128,
+    price2: u128
+) -> Observation {
+    return
+        Observation {
+            block_timestamp: block_timestamp,
+            num_of_observations: last.num_of_observations + 1,
+            price1_cumulative: last.price1_cumulative + price1,
+            price2_cumulative: last.price2_cumulative + price2,
+        };
+}
+
+impl Observation {
+    /// returns instance of Observation structure
+    pub fn new() -> Self {
+        return Self {
+            block_timestamp: 1,
+            num_of_observations: 0,
+            price1_cumulative: 0,
+            price2_cumulative: 0,
+        }
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    use near_sdk::test_utils::{accounts, VMContextBuilder};
+    use near_sdk::{testing_env, BlockHeight, MockedBlockchain};
+
+    fn init_blockchain() {
+        let context = VMContextBuilder::new();
+        testing_env!(context.build());
+    }
+
+    // returns twap with observation vector with timestamp [1, 2, 3, 4, 5, 6, 7, 9, 10]
+    fn get_twap(timestamp: u64) -> Twap {
+
+        let mut twap: Twap = Twap::new();
+        let mut current_idx = twap.initialize(timestamp, 1, 1);
+
+        let max_length = 10;
+        // fill all places
+        for i in 2..11 {
+            let timestamp = i;
+            current_idx = twap.write(
+                timestamp,
+                1, 1,
+                max_length
+            );
+        }
+
+        return twap;
+    }
+
+    #[test]
+    fn initialize_works() {
+        init_blockchain();
+
+        let mut twap: Twap = Twap::new();
+        let current_idx = twap.initialize(env::block_timestamp(), 1, 1);
```

Comment:
let's don't use `env::block_timestamp()` as well. Instead  - please use some numbers. Eg: `100`. It can be a const, defined in ` mod tests { ...`

Reply:
ok